data-pipeline-flink/user-cache-updater-2.0/src/main/scala/org/sunbird/dp/usercache/util/UserMetadataUpdater.scala

```diff
@@ -0,0 +1,254 @@
+          case _ => {
+            logger.info(s"Invalid event state name either it should be(Create/Created/Update/Updated) but found $name for ${event.mid()}")
+            metrics.incCounter(config.skipCount)
+            mutable.Map[String, AnyRef]()
+          }
+        }
+        if (!userData.isEmpty) {
+          val redisData = userData.filter{f => (null != f._2)}
+          val data = redisData.map(f => (f._1, f._2.toString))
+          val filteredData = mapAsJavaMap(data)
+          filteredData.values().removeAll(Collections.singleton({}))
+          dataCache.hmSet(id, filteredData)
+          metrics.incCounter(config.successCount)
+          metrics.incCounter(config.userCacheHit)
+        } else {
+          metrics.incCounter(config.skipCount)
+        }
+      }).getOrElse(metrics.incCounter(config.skipCount))
+    }).getOrElse(metrics.incCounter(config.skipCount))
+  }
+
+  /**
+    * When Edata.State is Create/Created then insert the user data into redis in string formate
+    */
+  def createAction(userId: String, event: Event, metrics: Metrics): mutable.Map[String, AnyRef] = {
+    val userData: mutable.Map[String, AnyRef] = mutable.Map[String, AnyRef]()
+    Option(event.getContextDataId(cDataType = "SignupType")).map(signInType => {
+      if (config.userSelfSignedInTypeList.contains(signInType)) {
+        userData.put(config.userSignInTypeKey, config.userSelfSignedKey)
+      }
+      if (config.userValidatedTypeList.contains(signInType)) {
+        userData.put(config.userSignInTypeKey, config.userValidatedKey)
+      }
+    }).orNull
+    userData
+  }
+
+  def updateAction(userId: String, event: Event, metrics: Metrics): mutable.Map[String, AnyRef] = {
+    val userCacheData: mutable.Map[String, String] = dataCache.hgetAllWithRetry(userId)
+
+    Option(event.getContextDataId(cDataType = "UserRole")).map(loginType => {
+      userCacheData.put(config.userLoginTypeKey, loginType)
+    })
+    val userMetaDataList = event.userMetaData()
+    if (!userMetaDataList.isEmpty() && userCacheData.contains(config.userSignInTypeKey) && ("Anonymous" != userCacheData.get(config.userSignInTypeKey))) {
+
+      // Get the user details from the cassandra table
+
+      val userDetails: mutable.Map[String, AnyRef] = userCacheData.++(extractUserMetaData(readFromCassandra(
+        keyspace = config.keySpace,
+        table = config.userTable,
+        QueryBuilder.eq("id", userId),
+        metrics)
+      ))
+      val updatedMap: mutable.Map[String, AnyRef] = userDetails.+=("iscustodianuser" -> isCustodianUser(userDetails, custodianRootOrgId).asInstanceOf[AnyRef])
+      val reportInfoMap: mutable.Map[String, AnyRef] = if (updatedMap.get("iscustodianuser").get == true) getCustodianUserInfo(updatedMap, metrics, userId) else getStateUserInfo(updatedMap, metrics, userId)
+
+      logger.info(s"User details ( $userId ) are fetched from the db's and updating the redis now.")
+      reportInfoMap
```

Comment:
Too many things happening here and little tough to follow. Please refactor to make it simpler to maintain. Can we do:
val userDetails = getUserDetails()
val userExternalDetails = getUserExternalDetails)
val updatedCache = userCacheData.++(userDetails).++(userExternalDetails)

Reply:
Done